<a href="https://colab.research.google.com/github/AnupamaSharma2000/yrbs_sexual_minority_analysis/blob/main/602_p3_anupamasharma.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Project 1 Redone


#Problem
    ○ Problem I aim to address throughout the course


1. Are LGBT students bullied more than their heterosexual peers?

2. Are LGBT children more prone to using illicit drugs than heterosexual children?

3. Do more kids identying as LBGT consider suicide compared to heterosexual peers?

4. Do LGBT kids feel more hopeless

5. Are teenager identifying as LGBT more likely to be forced to have sex.

6. Are prescription opioids misused more commonly among LGBT children?

7. Can I make a prediction here on if there are certain factors/experiences(ex: bullying, parental abuse) that might be cause to certain health risk behaviour. How will I make sure if my conclusions are not biased. How would a statistical analysis like look like?



The 2023 YRBS introduced questions on transgender identity for the first time. About 3% of high school students identify as transgender, with an additional 2% unsure about their transgender status. Transgender and questioning students report higher rates of violence victimization, unstable housing, and suicidal behaviors, and lower feelings of social connectedness at school




This dataset is personally interesting to me because, as someone who identifies as queer, I wanted to explore whether my experiences and behaviors are uniquely shaped by my identity. The inclusion of sexual and gender identity questions in this dataset motivated me to examine differences in experiences between sexual minority and majority youth.


I plan to analyze behavioral and experiential differences between sexual minority and majority youth. An investigable predictive variable could be mental health outcomes (e.g., suicidal ideation or attempts) based on sexual identity, experiences of bullying, and substance use patterns.


#Population
The focus of this study is on youth aged approximately 13-17 years, covering grades 9 through 12 in U.S. public and private schools, representing the national adolescent population.




#Variables

I have calculated two variables for sexual minority-
1. Sexual Minority Variable from sexual identity (q64)

2. Sexual Minority variable from sex and sex of sexual contacts (using q2 and q63)
Problem 1: Comparison between sexual minority and majority

Independent Variables

    (Sexual Minority Variables)
    Sexual Minority by Sexual Identity: q64 (Sexual identity)

    Sexual Minority by Sex and Sexual Contacts: q2 (Sex) and q63 (Sexual contact: who with)

Dependent Variables

    Behavioral and experiential differences related to bullying, substance use, and mental health outcomes.

    Including:

    Bullying: q23 (Treated badly), q24 (Bullied at school), q25 (Electronically bullied), q105 (Unfairly disciplined)

    Substance use: Tobacco (q31-q40), Alcohol (q41-q45), Illicit drugs (q46-q55), Pain medicine misuse (q49, q92)

    Mental health and suicide: q26-q30, q84, q106

    Sexual abuse and forced sex: q19-q22, q88, q90

    Suicidal behavior and hopelessness: covered under mental health and suicide variables

    School connectedness: q103 (Feel close to people at school)

       

Problem 2: Mental health outcomes in sexual minority(e.g. suicidal ideation or attempts) based on experiences of bullying, and substance use patterns.

Independent Variables
    Sexual Identity:

    q64: Sexual identity (defines sexual minority status)

    Experiences of Bullying:


    q24: Bullied at school

    q25: Electronically (cyber) bullied

    q105: Unfairly disciplined at school

    Substance Use Patterns:

    Tobacco: q31–q40 (e.g., ever smoked, days smoked)

    Alcohol: q41–q45 (age at first drank, days drank, binge drinking)

    Illicit drugs: q46–q55 (marijuana, cocaine, heroin, etc.)

    Pain medicine misuse: q49/q92

    Demographics: q1 (Age), q2 (Sex), q3 (Grade)
    Body metrics (could relate to mental health/substance use): q6 (Height), q7 (Weight)

    Parental factors:

    q89: Parent/adult insulted/put down

    q91: Parents/adults fought each other

    q99: Adult met basic needs

    q100: Lived with addict parent/guardian

    q101: Lived with mentally ill parent/guardian

    q102: Separated from parent due to incarceration

    

Dependent Variable
Mental Health Outcomes: I plan to merge these columns into one column

    q26: Felt sad/hopeless (2+ weeks)

    q27: Seriously considered suicide

    q28: Made suicide plan

    q29: Number of suicide attempts

    q30: Suicide attempt needed treatment

    q84: Mental health not good (30 days)

Confounding Variables

Extraneous variables that may affect relationships between independent and dependent variables,

Most of the variables that might act as confounders are already available in the data.


These are confounders in a way they affect the dependent and independent variabes so I will include these in the study(I wasn't initially planning to keep them in my analysis)



    Demographics:
   
    q1 (Age), q2 (Sex), q3 (Grade), race/ethnicity

    Body metrics: q6 (Height), q7 (Weight)

    Parental factors: q89 (Parent insulted), q91 (Parents fought), q99 (Adult met needs), q100 (Lived with addict parent), q101 (Lived with mentally ill parent), q102 (Separated from parent due to incarceration)


    Social and medical care: q80 (Social media use), q81 (Tested for HIV), q82 (STD test), q83 (Last dental visit)   

I will deal with them by:

Stratification: I will analyze groups separately or stratify data by key confounders like age group or sex to understand differential effects.

Matching: Where possible, will match participants on confounders to reduce bias
    
Replication: Validate findings across subpopulations and datasets.
    


#Dataset Chosen

    ○ Explain why this dataset is interesting to you

This is the Youth Risk Behavior Surveillance System (YRBSS) dataset for 2023. Currently this is the most well-documented data I could find for analysis on sexual minority adolescents.

- Developed - 1990

- Survey Frequency - Every 2 years

- Monitors priority health risk behaviors/experiences that contribute to the leading causes of death,
disability, and social problems among youth and adults in the United States.
-Includes national, state, territorial, tribal government, and local
school-based surveys of representative samples of ***9th through 12th grade  students.***
- Target population consists of all public, Catholic, and other
private school students.


Response Rates:

- School response rate – 155 of the 311 sampled schools participate
155/311 = 49.8% school response rate

- Student response rate – 20,386 of 28,308 sampled students submitted
questionnaires;

Usable  questionnaires - 20,103 (after after data editing).
20,836/23,308 = 71.0%

- Overall response rate = School response rate * Student response rate
49.8% * 71.0% = 35.4



Sample Description – 2023 National YRBS

- School-Level Selection
Included: All regular public, Catholic, and other private school students in grades 9–12.

- Coverage: 36 states, 3 tribal governments, 5 territories, and 21 local school districts.

- Excluded: Puerto Rico, the trust territories, and the Virgin Islands.

- Method: Schools were selected systematically with probability proportional to enrollment (grades 9–12) using a random start.

- Total sampled schools: 311.

- Class-Level Selection
Sampling frame: All classes in a required subject or all classes meeting during a specific period of the day, depending on school policy.

- Method: Systematic equal probability sampling with a random start was used to select classes within each participating school.

Age, Sex, Height, Weight and BMI are used to determine biologically implausible datapoint


    ○ What is in the dataset?

This survey dataset has the responses of students to 107 total questions of national  -
       [questionairre](https://drive.google.com/file/d/1vjy66o8hqAXymqot5U4Vf7uA52fsPaG-/view?usp=drive_link)





In [ ]:
# These are all the columns and their descriptions

{        #demographic
                                "q1": 'Age',
                                'q2': 'Sex',
                                'q3': 'Grade',

                                #body metric
                                'q6': 'Height(metres)',
                                'q7': 'Weight (kg)',

                                 #driving habits
                                "q8":  "Seat belt use (when riding)",
                                "q9":  "Times rode with drunk driver (30 days)",
                                "q10": "Times drove after drinking (30 days)",
                                "q11": "Days texted/emailed while driving (30 days)",


                                "q12": "Days carried weapon at school (30 days)",
                                "q13": "Days carried gun (12 months)",
                                "q14": "Days missed school for safety (30 days)",

                                 #sexual and physical violence

                                'q15': 'Threatened with weapon at School',
                                'q16': 'In a physical fight',
                                'q17': 'In a physical fight at school',
                                'q18': 'Seen physical violence',


                                 #bullying
                                'q23': 'Treated badly because of race/ethinicity',
                                'q24': 'Bullied at School',
                                'q25': 'Electronically(Cyber?) bullied',
                                'q105': 'Unfairly disciplined at school (12 months)',

                                 #mental health/ disability
                                "q26": "Felt sad/hopeless (2+ weeks)",
                                "q27": "Seriously considered suicide",
                                "q28": "Made suicide plan",
                                "q29": "Number of suicide attempts",
                                "q30": "Suicide attempt needed treatment",
                                "q84": "Mental health not good (30 days)",

                                 #Tobacco use
                                "q31": "Ever smoked cigarette",
                                "q32": "Age first smoked cigarette",
                                "q33": "Days smoked cigarettes (30 days)",
                                "q34": "Cigs/day on days smoked",
                                "q35": "Ever used electronic vapor product",
                                "q36": "Days used vapor product (30 days)",
                                "q37": "How obtained vapor products",
                                "q38": "Days used chewing tobacco (30 days)",
                                "q39": "Days smoked cigars (30 days)",

                                 #Alcohol use
                                "q41": "Age first drank alcohol",
                                "q42": "Days drank alcohol (30 days)",
                                "q43": "Days had 4+/5+ drinks in row (30 days)",
                                "q44": "Max drinks in a row (30 days)",                                           #not relevant for my analysis
                                "q45": "How obtained alcohol (30 days)",                                          #not relevant for my analysis

                                 #Marijuana use
                                "q46": "Times used marijuana (lifetime)",
                                "q47": "Age first tried marijuana",
                                "q48": "Times used marijuana (30 days)",


                                "q49": "Times misused pain medicine",
                                "q92": "Times misused pain medicine (30 days)",                                   #not intrested in recency, can remove(maybe?-revisit)

                                "q50": "Times used cocaine",
                                "q51": "Times used inhalants",
                                "q52": "Times used heroin",
                                "q53": "Times used methamphetamines",
                                "q54": "Times used ecstasy",
                                "q55": "Times injected illegal drugs",


                                "q93": "Times used hallucinogens (lifetime)",

                                "q40": "Tried to quit all tobacco (12 months)",

                                 #sexual activity and identity
                                "q56": "Ever had sexual intercourse",
                                "q57": "Age at first intercourse",
                                "q58": "Lifetime sexual partners",
                                "q59": "Sexual partners (past 3 months)",

                                "q60": "Used alcohol/drugs before last sex",

                                "q61": "Used condom at last sex",
                                "q62": "Pregnancy prevention method at last sex",

                                 #Sexual minority
                                "q63": "Sexual contact: who with (lifetime)",
                                "q64": "Sexual identity",
                                "q65": "Transgender identity",

                                "q94": "Asked for verbal consent at last sexual encounter",


                                 #weight, eating habits, sleep
                                "q66": "Describe your weight",
                                "q67": "Current weight control effort",
                                "q68": "Drank 100% fruit juice (7 days)",
                                "q69": "Ate fruit (7 days)",
                                "q70": "Ate green salad (7 days)",
                                "q71": "Ate potatoes (7 days)",
                                "q72": "Ate carrots (7 days)",
                                "q73": "Ate vegetables (7 days)",
                                "q74": "Drank soda/pop (7 days)",
                                "q75": "Days ate breakfast (7 days)",
                                "q85": "Sleep hours (school night)",
                                "q86": "Slept usually where (30 days)",

                                 #Exercise
                                "q76": "Days active 60+ min (7 days)",
                                "q77": "PE class days (week)",
                                "q78": "Sports teams played (12 months)",


                                 #Injury
                                "q79": "Concussions from sports/activity (12 months)",
                                "q98": "Times had sunburn (12 months)",


                                 #Social
                                "q80": "Social media use frequency",

                                 #Medical care
                                "q81": "Tested for HIV (ever)",                            #the question asks if tested not tested positive
                                "q82": "STD test (12 months)",
                                "q83": "Last dental visit",


                                #Experiences physical or sexual abuse
                                'q19': 'Physically Forced to have sex',                               #check if mental coercion also included later
                                'q20': 'Times Forced to do sexual things',
                                'q21': 'Forced by a partner(loosely) to do sexual things',             #while doing analysis for person identifying as LGBT note we dont know which gender they are dating or smth
                                'q22': 'Physically hurt by partner',
                                "q88": "Sexual abuse by older adult",
                                "q90": "Parent/adult physical abuse (lifetime)",


                                #Parental factors
                                "q89": "Parent/adult insulted/put down (lifetime)",
                                "q91": "Parents/adults fought each other (lifetime)",
                                "q99": "Adult met basic needs (lifetime)",
                                "q100": "Lived with addict parent/guardian",
                                "q101": "Lived with mentally ill parent/guardian",
                                "q102": "Separated from parent because of incarceration",
                                "q104": "An Adult knows whereabouts",



                                #School
                                "q87": "Grades in school (12 months)",
                                "q103": "Feel close to people at school",



                                "q107": "English speaking ability"


}


YRBSS monitors six categories of priority health-risk behaviors
among youths and young adults:

1) behaviors that contribute to unintentional injuries and violence; 2) sexual behaviors that
contribute to human immunodeficiency virus (HIV) infection, other sexually transmitted diseases, and unintended pregnancy;
3) tobacco use;
4) alcohol and other drug use;
5) unhealthy dietary behaviors; and
6) physical inactivity. In addition, YRBSS
monitors the prevalence of obesity and asthma among this population.

They first picked schools so that big schools got picked more often than small ones (fair representation). Then inside each school, they randomly picked classes so students had an equal shot at being in the study. This keeps the survey cost-effective yet still represents the bigger population accurately.


 A weighting factor was
applied to each student record to adjust for nonresponse and the
oversampling of Black and Hispanic students in the sample.

#Hypothesis:
I hypothesize that my analysis will reveal:


    LGBT students experience higher rates of bullying, substance use, and mental health challenges compared to heterosexual peers.

    Sexual minority youth will report more suicidal ideation, suicide planning, and attempts.

    Illicit drug use and misuse of prescription opioids will be more common among LGBT adolescents.

    Experiences of bullying and discrimination will mediate the relationship between sexual minority status and adverse mental health outcomes.

#Data Collection
    ○ How will I get the data?
I will collect data through school based surveys administered to a representative sample of students in grades 9-12, similar to the YRBS approach. Surveys will be distributed in classrooms during school hours to maximize participation and representativeness.

    ○ How will I ensure representativeness when using a sample?
Use a three-stage cluster sampling design:

Systematically sample schools stratified by region and size.

Systematically select classes from these schools across grade levels.

Include all students in selected classes who consent/assent to participate.

Apply weighting factors to adjust for oversampling of minority and high-risk groups, nonresponse bias, and to align sample demographics with national population benchmarks.

Conduct regular response rate monitoring at the school and student levels to identify and mitigate participation gaps.


    ○ Method to collect the data?

Method to collect the data?
Surveys will be self-administered anonymous questionnaires completed during class time, enhancing candid responses.


Employ strict data editing protocols for valid responses:

Range and logical consistency checks

Invalid/multiple responses coded as missing

Height and weight plausibility checks using established conversion formulas and biological limits

Implement confidentiality protections to encourage truthful reporting, particularly for sensitive questions

In [ ]:
import pandas as pd

raw_data= pd.read_csv("/content/drive/MyDrive/pods_project1/new_data_project_2/XXHq.csv")

/tmp/ipython-input-3628982452.py:3: DtypeWarning: Columns (2) have mixed types. Specify dtype option on import or set low_memory=False.
  raw_data= pd.read_csv("/content/drive/MyDrive/pods_project1/new_data_project_2/XXHq.csv")


In [ ]:

raw_data.head()

site  raceeth q6orig q7orig  record  orig_rec   q1   q2   q3   q4  ... q102  \
0   XX      NaN    505    180     1.0       NaN  3.0  1.0  1.0  NaN  ...  2.0   
1   XX      5.0    N N    233     2.0       NaN  4.0  2.0  1.0  2.0  ...  2.0   
2   XX      5.0    506    165     3.0       NaN  5.0  2.0  3.0  2.0  ...  1.0   
3   XX      5.0    N N    105     4.0       NaN  6.0  1.0  2.0  2.0  ...  2.0   
4   XX      5.0    601    125     5.0       NaN  3.0  2.0  1.0  2.0  ...  2.0   

   q103  q104  q105  q106  q107  BMIPCT  weight  stratum      psu  
0   3.0   3.0   2.0   2.0   2.0   97.08    0.86    103.0  16294.0  
1   2.0   5.0   2.0   1.0   2.0     NaN    0.89    103.0  16294.0  
2   2.0   4.0   2.0   2.0   1.0   92.26    0.50    103.0  16294.0  
3   4.0   5.0   2.0   1.0   1.0     NaN    1.17    103.0  16294.0  
4   2.0   5.0   2.0   2.0   1.0    7.56    0.89    103.0  16294.0  

[5 rows x 117 columns]

In [ ]:
raw_raceeth_mapping_data = [
    {"Category": "Ethnicity Race Raceeth", "Ethnicity": "B", "Race": "A","Value":1, "Label": "American Indian/Alaskan Native"},
    {"Category": "Ethnicity Race Raceeth", "Ethnicity": "B", "Race": "B","Value":2, "Label": "Asian"},
    {"Category": "Ethnicity Race Raceeth", "Ethnicity": "B", "Race": "C","Value":3, "Label": "Black or African American"},
    {"Category": "Ethnicity Race Raceeth", "Ethnicity": "B", "Race": "D","Value":4, "Label": "Native Hawaiian or Other Pacific Islander"},
    {"Category": "Ethnicity Race Raceeth", "Ethnicity": "B", "Race": "E","Value":5, "Label": "White"},
    {"Category": "Ethnicity Race Raceeth", "Ethnicity": "A", "Race": "Missing","Value":6, "Label": "Hispanic/Latino"},
    {"Category": "Ethnicity Race Raceeth", "Ethnicity": "A", "Race": "1 or more responses","Value":7, "Label": "Multiple– Hispanic/Latino"},
    {"Category": "Ethnicity Race Raceeth", "Ethnicity": "B", "Race": "2 or more responses","Value":8, "Label": "Multiple– Non-Hispanic/Latino"},
    {"Category": "Ethnicity Race Raceeth", "Ethnicity": "B", "Race": "Missing","Value":pd.NA, "Label": "Missing"},
    {"Category": "Ethnicity Race Raceeth", "Ethnicity": "Missing", "Race": "Missing or Any response","Value":pd.NA, "Label": "Missing"},
    {"Category": "Ethnicity Race Raceeth", "Ethnicity": "Out of range", "Race": "Any response","Value":pd.NA, "Label": "Missing"},
    {"Category": "Ethnicity Race Raceeth", "Ethnicity": "Any response", "Race": "Out of range","Value":pd.NA, "Label": "Missing"},
]
raceeth_mapping_df = pd.DataFrame(raw_raceeth_mapping_data)

In [ ]:
raceeth_mapping_df

Category     Ethnicity                     Race Value  \
0   Ethnicity Race Raceeth             B                        A     1   
1   Ethnicity Race Raceeth             B                        B     2   
2   Ethnicity Race Raceeth             B                        C     3   
3   Ethnicity Race Raceeth             B                        D     4   
4   Ethnicity Race Raceeth             B                        E     5   
5   Ethnicity Race Raceeth             A                  Missing     6   
6   Ethnicity Race Raceeth             A      1 or more responses     7   
7   Ethnicity Race Raceeth             B      2 or more responses     8   
8   Ethnicity Race Raceeth             B                  Missing  <NA>   
9   Ethnicity Race Raceeth       Missing  Missing or Any response  <NA>   
10  Ethnicity Race Raceeth  Out of range             Any response  <NA>   
11  Ethnicity Race Raceeth  Any response             Out of range  <NA>   

                                        Label  
0              American Indian/Alaskan Native  
1                                       Asian  
2                   Black or African American  
3   Native Hawaiian or Other Pacific Islander  
4                                       White  
5                             Hispanic/Latino  
6                   Multiple– Hispanic/Latino  
7               Multiple– Non-Hispanic/Latino  
8                                     Missing  
9                                     Missing  
10                                    Missing  
11                                    Missing

#Project 2

##Preliminary information about the dataset

Shape of the data


In [ ]:
##Shape
raw_data.shape


(20103, 117)

Datatypes in the data



In [ ]:
raw_data.dtypes

site        object
raceeth    float64
q6orig      object
q7orig      object
record     float64
            ...   
q107       float64
BMIPCT     float64
weight     float64
stratum    float64
psu        float64
Length: 117, dtype: object

In [ ]:
#can't to show datatypes of each column since 117 columns is a lot, so printing counts of each datatype
datatype_counts = raw_data.dtypes.value_counts()
print(datatype_counts)

float64    113
object       4
Name: count, dtype: int64


All variables/column in the data

In [ ]:
columns_list = raw_data.columns.tolist()
print(columns_list)

['site', 'raceeth', 'q6orig', 'q7orig', 'record', 'orig_rec', 'q1', 'q2', 'q3', 'q4', 'q5', 'q6', 'q7', 'q8', 'q9', 'q10', 'q11', 'q12', 'q13', 'q14', 'q15', 'q16', 'q17', 'q18', 'q19', 'q20', 'q21', 'q22', 'q23', 'q24', 'q25', 'q26', 'q27', 'q28', 'q29', 'q30', 'q31', 'q32', 'q33', 'q34', 'q35', 'q36', 'q37', 'q38', 'q39', 'q40', 'q41', 'q42', 'q43', 'q44', 'q45', 'q46', 'q47', 'q48', 'q49', 'q50', 'q51', 'q52', 'q53', 'q54', 'q55', 'q56', 'q57', 'q58', 'q59', 'q60', 'q61', 'q62', 'q63', 'q64', 'q65', 'q66', 'q67', 'q68', 'q69', 'q70', 'q71', 'q72', 'q73', 'q74', 'q75', 'q76', 'q77', 'q78', 'q79', 'q80', 'q81', 'q82', 'q83', 'q84', 'q85', 'q86', 'q87', 'q88', 'q89', 'q90', 'q91', 'q92', 'q93', 'q94', 'q95', 'q96', 'q97', 'q98', 'q99', 'q100', 'q101', 'q102', 'q103', 'q104', 'q105', 'q106', 'q107', 'BMIPCT', 'weight', 'stratum', 'psu']


In [ ]:
raw_data.describe()

raceeth        record  orig_rec            q1            q2  \
count  19733.000000  20103.000000       0.0  20005.000000  19945.000000   
mean       5.010845  10140.386161       NaN      4.892677      1.504437   
std        1.832258   5863.836462       NaN      1.218455      0.499993   
min        1.000000      1.000000       NaN      1.000000      1.000000   
25%        5.000000   5062.500000       NaN      4.000000      1.000000   
50%        5.000000  10129.000000       NaN      5.000000      2.000000   
75%        6.000000  15203.500000       NaN      6.000000      2.000000   
max        8.000000  20386.000000       NaN      7.000000      2.000000   

                 q3            q4            q6            q7            q8  \
count  19910.000000  19849.000000  17814.000000  17814.000000  15071.000000   
mean       2.361477      1.798630      1.696357     68.680140      4.310796   
std        1.101724      0.401034      0.102717     18.466386      0.975215   
min        1.000000      1.000000      1.320000     31.300000      1.000000   
25%        1.000000      2.000000      1.630000     55.790000      4.000000   
50%        2.000000      2.000000      1.700000     64.410000      5.000000   
75%        3.000000      2.000000      1.780000     77.110000      5.000000   
max        5.000000      2.000000      2.030000    180.990000      5.000000   

       ...          q102          q103          q104          q105  \
count  ...  13342.000000  11177.000000  10850.000000  10583.000000   
mean   ...      1.823115      2.618323      4.205253      1.794860   
std    ...      0.381586      1.144287      0.999396      0.403823   
min    ...      1.000000      1.000000      1.000000      1.000000   
25%    ...      2.000000      2.000000      4.000000      2.000000   
50%    ...      2.000000      2.000000      4.000000      2.000000   
75%    ...      2.000000      3.000000      5.000000      2.000000   
max    ...      2.000000      5.000000      5.000000      2.000000   

               q106          q107        BMIPCT        weight       stratum  \
count  10946.000000  10795.000000  1.781400e+04  20103.000000  20103.000000   
mean       1.519368      1.210283  6.337486e+01      0.995512    228.370243   
std        0.499648      0.494050  2.931234e+01      1.270584     73.287806   
min        1.000000      1.000000  1.597498e-08      0.002300    101.000000   
25%        1.000000      1.000000  4.149000e+01      0.150000    201.000000   
50%        2.000000      1.000000  6.887000e+01      0.410000    212.000000   
75%        2.000000      1.000000  9.017000e+01      1.350000    300.000000   
max        2.000000      4.000000  9.996000e+01     22.540000    300.000000   

                psu  
count   20103.00000  
mean   414571.68925  
std    235438.54837  
min     16294.00000  
25%    214944.00000  
50%    401089.00000  
75%    573255.00000  
max    788430.00000  

[8 rows x 113 columns]

In [ ]:
num_unique_elements = raw_data.drop_duplicates().shape[0]
num_unique_elements                                                    #20,103 questionnaires were usable after data editing (out of 23,308) = 71.0% , page 7 in user guide for data edits

20103

It makes sense that our total unique number of rows is equal to the number of usable questionnaires in the survey data as each row is a particular student

##Specific data needs for my project

There are too many columns (117), I am removing sum columns based on the my current understanding that they won't be useful for my analysis


In [ ]:
#dropping some columns that are not useful for my analysis

columns_to_drop = ["orig_rec","q8", "q9", "q10", "q11","q12","q37","q40","q45","q44","q66", "q67", "q68", "q69", "q70", "q71", "q72",
                   "q73", "q74", "q75", "q76","q77","q78", "q79","q92","q98","q102","q107"]
raw_data = raw_data.drop(columns=columns_to_drop)


In [ ]:
columns_list = raw_data.columns.tolist()
print(columns_list)

['site', 'raceeth', 'q6orig', 'q7orig', 'record', 'orig_rec', 'q1', 'q2', 'q3', 'q4', 'q5', 'q6', 'q7', 'q13', 'q14', 'q15', 'q16', 'q17', 'q18', 'q19', 'q20', 'q21', 'q22', 'q23', 'q24', 'q25', 'q26', 'q27', 'q28', 'q29', 'q30', 'q31', 'q32', 'q33', 'q34', 'q35', 'q36', 'q38', 'q39', 'q41', 'q42', 'q43', 'q46', 'q47', 'q48', 'q49', 'q50', 'q51', 'q52', 'q53', 'q54', 'q55', 'q56', 'q57', 'q58', 'q59', 'q60', 'q61', 'q62', 'q63', 'q64', 'q65', 'q80', 'q81', 'q82', 'q83', 'q84', 'q85', 'q86', 'q87', 'q88', 'q89', 'q90', 'q91', 'q93', 'q94', 'q95', 'q96', 'q97', 'q99', 'q100', 'q101', 'q103', 'q104', 'q105', 'q106', 'BMIPCT', 'weight', 'stratum', 'psu']


    ○ Multiple datasets that need to be merged?
I am only using one table named XXHq.csv.

There is one more table named XXHqn.csv which contains dichotomous variables but I plan to calculate some of them on my own.

Example of dichotomous variable (from the codebook):

      Numerator: Students who answered B for Q30
      Denominator: Students who answered A, B, or C for Q30
      Summary text: Percentage of students who had a suicide attempt that resulted in an injury, poisoning, or
      overdose that had to be treated by a doctor or nurse (during the 12 months before the
      survey)
      QN variable label: Had a suicide attempt that resulted in an injury, poisoning, or overdose that had to be
      treated by a doctor or nurse
      Native direction: Negative

Multiple datasets don't need to be merged

##Potential issues in the data

1. There are no duplicates in my dataset, since it's a survey dataset and each row entry represents a usable filled questionnaire by a student.

2. Missing or null values

In [ ]:
missing_df = raw_data.isnull().sum().reset_index()
missing_df.columns = ['Column', 'Missing Values']
missing_df = missing_df.sort_values(by='Missing Values', ascending=False)
print(missing_df.to_string(index=False))

  Column  Missing Values
orig_rec           20103
    q105            9520
    q104            9253
    q106            9157
     q93            8932
    q103            8926
     q97            8906
     q94            8770
     q13            8729
     q16            8451
     q95            8194
     q34            7946
     q82            7325
    q100            6876
    q101            6798
     q96            6519
     q91            6350
     q88            6280
     q90            6186
     q89            6183
     q51            5877
     q86            5562
     q18            5526
     q30            5468
     q55            5006
     q80            4900
     q63            4722
     q43            4662
     q81            4584
     q21            4504
     q84            4398
     q99            4353
     q20            4351
     q56            3982
     q60            3655
     q54            3608
     q87            3568
     q46            3500
     q32            3283


Three types of missing data:
1. Missing completely at random, MCAR

        When the missingness is unrelated to any variable or the value itself — typically caused by data-entry or recording errors, in this case participant might just be lazy to fill the survey.
2. Missing at random, MAR

       When missingness depends on other observed variables, not on the value itself

3. Missing not at random, MNAR

       When missingness depends on the value itself — common with sensitive or stigmatized topics.

Some missing data maybe because of Data edits done - see page 7 of guide

MCAR

    q1 (Age), q2 (Sex), q3 (Grade):
    might be missing when respondents skip demographic questions or their response is illegible or data are lost during entry.

    q8–q11 (Driving behaviors) and q13–q14 (Weapons/Safety):
    might be missing when some respondents may have skipped if not applicable or due to random omission.

    q6 (Height), q7 (Weight), q84 (Mental health not good):
    might be missing due to random scanning or transcription errors, not linked to personal traits.

    q87 (Grades), q103 (Feel close to people at school):
    Might be missing from random administrative or recording issues.

MAR

    q31–q45, q46–q55 (Substance use and sexual activity):
    Nonresponse depends on age, grade, or site but not directly on the missing variable’s true value.

    q56–q65 (Sexual activity/identity):
    Missing more likely if respondent skips related identity or orientation questions.

    q85–q86 (Sleep):
    Missing associated with housing instability patterns (linked to demographics or stress).

    q82–q83 (STD test, dental visit):
    Missingness relates to socioeconomic status or access to healthcare variables.


MNAR

Fear of repercussions


    q105 Since it asks about unfair discipline at school,students filling out the survey in a classrooms setting might omit or avoid answering truthfully due to fear of repercussions
    This missingness depends on the unreported value itself (the experience of unfair discipline)

Sensitive or traumatic experiences:

    q26–q30 (Mental health/suicide variables):
    Respondents with certain demographic characteristics (e.g., male, younger, lower grade) are less likely to answer due to stigma.

    q19–q22 (Sexual assault, partner violence)

    q88–q91 (Parental and sexual abuse)
    Respondents often skip these because disclosure depends on the experience itself.

High-stigma mental health and substance questions:

    q93 (Hallucinogen use), q55 (Injected drug use) often omitted by those who have engaged in high-risk behavior.

    q94 (Asked for verbal consent at last sex):
    Sensitive — more likely skipped by those who did not ask for consent.

    q90–q101 (Family dysfunction):
    might be missing due to unwillingness to report negative family environments.

    q102 (Separated from parent for incarceration):
    Nonresponse likely higher among those directly affected.



3. Formatting:
The variables are named as q1, q2, etc. It was time consuming to go look at the codebook for every question variable


Note - Incorrect data -participant lying on survey maybe(even though data edits sets most inconsistencies with missing values this can still be really useful in my data but remember for project 3)

There are no types inconsistencies.

There are no misspellings or mistypes but the 2 columns q6orig and q7orig are only mentioned once in the codebook but their description is not there.


Outliers:

    Since this is a survey dataset with certain specific options to choose from there are no outliers.
    The only possibility of outliers could be age, weight, and height but these inconsistencies were already solved by the data provider.



##Any reorganization needed?

1. Renaming these columns:

                                "q1": 'Age',
                                'q2': 'Sex',
                                'q3': 'Grade',
                                 
                                #body metric
                                'q6': 'Height(metres)',
                                'q7': 'Weight (kg)',

In [ ]:
raw_data = raw_data.rename(columns={
    'q1': 'Age',
    'q2': 'Sex',
    'q3': 'Grade',
    'q6': 'Height_meters',
    'q7': 'Weight_kg'
})

Based on my analysis my data doesn't have outliers and duplicates.

The only problem I have is missing values, but my data is big(20103 rows) so even with highest missing value column(9520) I have more than 10000 rows of data.

For MAR:

Since this is a categorical survey data, missing can just be a new category.

I would drop some datapoint according to my analysis as I gain more experience about this dataset(for example I can drop commmon missing point from dependent variables later)
 But droppping rows right now does not make sense to me as for a particular row some values might be missing but some might not and can be helpful.


MNAR
The variable that has the most missing values(105), according to my understanding seems to be MNAR. But this variable is not significant for my problem as of right now.

 My most important variables - q63, q64 - to calculate sexual minority - to only have close to 4000 and 2000 missing values which is not bad.

Here this is Non-Ignorable (NI) missing data, informative missing data



Note -  My bigger problem is incorrect data as participants could simply lie about their sexual identity





In [ ]:
raw_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20103 entries, 0 to 20102
Columns: 117 entries, site to psu
dtypes: float64(113), object(4)
memory usage: 17.9+ MB


In [ ]:
raw_data.describe()

raceeth        record  orig_rec            q1            q2  \
count  19733.000000  20103.000000       0.0  20005.000000  19945.000000   
mean       5.010845  10140.386161       NaN      4.892677      1.504437   
std        1.832258   5863.836462       NaN      1.218455      0.499993   
min        1.000000      1.000000       NaN      1.000000      1.000000   
25%        5.000000   5062.500000       NaN      4.000000      1.000000   
50%        5.000000  10129.000000       NaN      5.000000      2.000000   
75%        6.000000  15203.500000       NaN      6.000000      2.000000   
max        8.000000  20386.000000       NaN      7.000000      2.000000   

                 q3            q4            q6            q7            q8  \
count  19910.000000  19849.000000  17814.000000  17814.000000  15071.000000   
mean       2.361477      1.798630      1.696357     68.680140      4.310796   
std        1.101724      0.401034      0.102717     18.466386      0.975215   
min        1.000000      1.000000      1.320000     31.300000      1.000000   
25%        1.000000      2.000000      1.630000     55.790000      4.000000   
50%        2.000000      2.000000      1.700000     64.410000      5.000000   
75%        3.000000      2.000000      1.780000     77.110000      5.000000   
max        5.000000      2.000000      2.030000    180.990000      5.000000   

       ...          q102          q103          q104          q105  \
count  ...  13342.000000  11177.000000  10850.000000  10583.000000   
mean   ...      1.823115      2.618323      4.205253      1.794860   
std    ...      0.381586      1.144287      0.999396      0.403823   
min    ...      1.000000      1.000000      1.000000      1.000000   
25%    ...      2.000000      2.000000      4.000000      2.000000   
50%    ...      2.000000      2.000000      4.000000      2.000000   
75%    ...      2.000000      3.000000      5.000000      2.000000   
max    ...      2.000000      5.000000      5.000000      2.000000   

               q106          q107        BMIPCT        weight       stratum  \
count  10946.000000  10795.000000  1.781400e+04  20103.000000  20103.000000   
mean       1.519368      1.210283  6.337486e+01      0.995512    228.370243   
std        0.499648      0.494050  2.931234e+01      1.270584     73.287806   
min        1.000000      1.000000  1.597498e-08      0.002300    101.000000   
25%        1.000000      1.000000  4.149000e+01      0.150000    201.000000   
50%        2.000000      1.000000  6.887000e+01      0.410000    212.000000   
75%        2.000000      1.000000  9.017000e+01      1.350000    300.000000   
max        2.000000      4.000000  9.996000e+01     22.540000    300.000000   

                psu  
count   20103.00000  
mean   414571.68925  
std    235438.54837  
min     16294.00000  
25%    214944.00000  
50%    401089.00000  
75%    573255.00000  
max    788430.00000  

[8 rows x 113 columns]

In [ ]:
# extracted_raw_data = raw_data['site','raceeth','q1','q2','q3','q6','q7','q10','q11','q14','q15','q19', 'q21', 'q24','q25','q26','q27','q28','q29','q30', 'q31','q84','q105','q106']


Tags = ['Identifier', 'Factors/Experiences', 'Behaviours', 'outcomes']

feature_descriptions ={        #demographic
                                "q1": 'Age',
                                'q2': 'Sex',
                                'q3': 'Grade',

                                #body metric
                                'q6': 'Height(metres)',
                                'q7': 'Weight (kg)',

                                 #driving habits
                                "q8":  "Seat belt use (when riding)",
                                "q9":  "Times rode with drunk driver (30 days)",
                                "q10": "Times drove after drinking (30 days)",
                                "q11": "Days texted/emailed while driving (30 days)",


                                "q12": "Days carried weapon at school (30 days)",
                                "q13": "Days carried gun (12 months)",
                                "q14": "Days missed school for safety (30 days)",

                                 #sexual and physical violence

                                'q15': 'Threatened with weapon at School',
                                'q16': 'In a physical fight',
                                'q17': 'In a physical fight at school',
                                'q18': 'Seen physical violence',


                                 #bullying
                                'q23': 'Treated badly because of race/ethinicity',
                                'q24': 'Bullied at School',
                                'q25': 'Electronically(Cyber?) bullied',
                                'q105': 'Unfairly disciplined at school (12 months)',

                                 #mental health/ disability
                                "q26": "Felt sad/hopeless (2+ weeks)",
                                "q27": "Seriously considered suicide",
                                "q28": "Made suicide plan",
                                "q29": "Number of suicide attempts",
                                "q30": "Suicide attempt needed treatment",
                                "q84": "Mental health not good (30 days)",

                                 #Tobacco use
                                "q31": "Ever smoked cigarette",
                                "q32": "Age first smoked cigarette",
                                "q33": "Days smoked cigarettes (30 days)",
                                "q34": "Cigs/day on days smoked",
                                "q35": "Ever used electronic vapor product",
                                "q36": "Days used vapor product (30 days)",
                                "q37": "How obtained vapor products",
                                "q38": "Days used chewing tobacco (30 days)",
                                "q39": "Days smoked cigars (30 days)",

                                 #Alcohol use
                                "q41": "Age first drank alcohol",
                                "q42": "Days drank alcohol (30 days)",
                                "q43": "Days had 4+/5+ drinks in row (30 days)",
                                "q44": "Max drinks in a row (30 days)",                                           #not relevant for my analysis
                                "q45": "How obtained alcohol (30 days)",                                          #not relevant for my analysis

                                 #Marijuana use
                                "q46": "Times used marijuana (lifetime)",
                                "q47": "Age first tried marijuana",
                                "q48": "Times used marijuana (30 days)",


                                "q49": "Times misused pain medicine",
                                "q92": "Times misused pain medicine (30 days)",                                   #not intrested in recency, can remove(maybe?-revisit)

                                "q50": "Times used cocaine",
                                "q51": "Times used inhalants",
                                "q52": "Times used heroin",
                                "q53": "Times used methamphetamines",
                                "q54": "Times used ecstasy",
                                "q55": "Times injected illegal drugs",


                                "q93": "Times used hallucinogens (lifetime)",

                                "q40": "Tried to quit all tobacco (12 months)",

                                 #sexual activity and identity
                                "q56": "Ever had sexual intercourse",
                                "q57": "Age at first intercourse",
                                "q58": "Lifetime sexual partners",
                                "q59": "Sexual partners (past 3 months)",

                                "q60": "Used alcohol/drugs before last sex",

                                "q61": "Used condom at last sex",
                                "q62": "Pregnancy prevention method at last sex",

                                 #Sexual minority
                                "q63": "Sexual contact: who with (lifetime)",
                                "q64": "Sexual identity",
                                "q65": "Transgender identity",

                                "q94": "Asked for verbal consent at last sexual encounter",


                                 #weight, eating habits, sleep
                                "q66": "Describe your weight",
                                "q67": "Current weight control effort",
                                "q68": "Drank 100% fruit juice (7 days)",
                                "q69": "Ate fruit (7 days)",
                                "q70": "Ate green salad (7 days)",
                                "q71": "Ate potatoes (7 days)",
                                "q72": "Ate carrots (7 days)",
                                "q73": "Ate vegetables (7 days)",
                                "q74": "Drank soda/pop (7 days)",
                                "q75": "Days ate breakfast (7 days)",
                                "q85": "Sleep hours (school night)",
                                "q86": "Slept usually where (30 days)",

                                 #Exercise
                                "q76": "Days active 60+ min (7 days)",
                                "q77": "PE class days (week)",
                                "q78": "Sports teams played (12 months)",


                                 #Injury
                                "q79": "Concussions from sports/activity (12 months)",
                                "q98": "Times had sunburn (12 months)",


                                 #Social
                                "q80": "Social media use frequency",

                                 #Medical care
                                "q81": "Tested for HIV (ever)",                            #the question asks if tested not tested positive
                                "q82": "STD test (12 months)",
                                "q83": "Last dental visit",


                                #Experiences physical or sexual abuse
                                'q19': 'Physically Forced to have sex',                               #check if mental coercion also included later
                                'q20': 'Times Forced to do sexual things',
                                'q21': 'Forced by a partner(loosely) to do sexual things',             #while doing analysis for person identifying as LGBT note we dont know which gender they are dating or smth
                                'q22': 'Physically hurt by partner',
                                "q88": "Sexual abuse by older adult",
                                "q90": "Parent/adult physical abuse (lifetime)",


                                #Parental factors
                                "q89": "Parent/adult insulted/put down (lifetime)",
                                "q91": "Parents/adults fought each other (lifetime)",
                                "q99": "Adult met basic needs (lifetime)",
                                "q100": "Lived with addict parent/guardian",
                                "q101": "Lived with mentally ill parent/guardian",
                                "q102": "Separated from parent because of incarceration",
                                "q104": "An Adult knows whereabouts",



                                #School
                                "q87": "Grades in school (12 months)",
                                "q103": "Feel close to people at school",



                                "q107": "English speaking ability"


}


In [ ]:
import pandas as pd

data = [
    # Demographics
    {'Variable Name': 'q1', 'Meaning': 'Age', 'Tag': 'Demographics'},
    {'Variable Name': 'q2', 'Meaning': 'Sex', 'Tag': 'Demographics'},
    {'Variable Name': 'q3', 'Meaning': 'Grade', 'Tag': 'Demographics'},

    # Body Metrics
    {'Variable Name': 'q6', 'Meaning': 'Height(metres)', 'Tag': 'Body Metrics'},
    {'Variable Name': 'q7', 'Meaning': 'Weight (kg)', 'Tag': 'Body Metrics'},

    # Driving Behaviors
    {'Variable Name': 'q8', 'Meaning': 'Seat belt use (when riding)', 'Tag': 'Driving Behaviors'},
    {'Variable Name': 'q9', 'Meaning': 'Times rode with drunk driver (30 days)', 'Tag': 'Driving Behaviors'},
    {'Variable Name': 'q10', 'Meaning': 'Times drove after drinking (30 days)', 'Tag': 'Driving Behaviors'},
    {'Variable Name': 'q11', 'Meaning': 'Days texted/emailed while driving (30 days)', 'Tag': 'Driving Behaviors'},

    # School Safety and Weapon Carrying
    {'Variable Name': 'q12', 'Meaning': 'Days carried weapon at school (30 days)', 'Tag': 'Weapon Carrying'},
    {'Variable Name': 'q13', 'Meaning': 'Days carried gun (12 months)', 'Tag': 'Weapon Carrying'},
    {'Variable Name': 'q14', 'Meaning': 'Days missed school for safety (30 days)', 'Tag': 'School Safety'},
    {'Variable Name': 'q15', 'Meaning': 'Threatened with weapon at School', 'Tag': 'School Safety & Weapon Carrying'},

    # Violence Exposure and Experience
    {'Variable Name': 'q16', 'Meaning': 'In a physical fight', 'Tag': 'Violence Exposure and Experience'},
    {'Variable Name': 'q17', 'Meaning': 'In a physical fight at school', 'Tag': 'Violence Exposure and Experience'},
    {'Variable Name': 'q18', 'Meaning': 'Seen physical violence', 'Tag': 'Violence Exposure and Experience'},

    # Bullying, Discrimination, Discipline
    {'Variable Name': 'q23', 'Meaning': 'Treated badly because of race/ethnicity', 'Tag': 'Bullying, Discrimination, Discipline'},
    {'Variable Name': 'q24', 'Meaning': 'Bullied at School', 'Tag': 'Bullying, Discrimination, Discipline'},
    {'Variable Name': 'q25', 'Meaning': 'Electronically(Cyber?) bullied', 'Tag': 'Bullying, Discrimination, Discipline'},
    {'Variable Name': 'q105', 'Meaning': 'Unfairly disciplined at school (12 months)', 'Tag': 'Bullying, Discrimination, Discipline'},

    # Mental Health & Suicide
    {'Variable Name': 'q26', 'Meaning': 'Felt sad/hopeless (2+ weeks)', 'Tag': 'Mental Health & Suicide'},
    {'Variable Name': 'q27', 'Meaning': 'Seriously considered suicide', 'Tag': 'Mental Health & Suicide'},
    {'Variable Name': 'q28', 'Meaning': 'Made suicide plan', 'Tag': 'Mental Health & Suicide'},
    {'Variable Name': 'q29', 'Meaning': 'Number of suicide attempts', 'Tag': 'Mental Health & Suicide'},
    {'Variable Name': 'q30', 'Meaning': 'Suicide attempt needed treatment', 'Tag': 'Mental Health & Suicide'},
    {'Variable Name': 'q84', 'Meaning': 'Mental health not good (30 days)', 'Tag': 'Mental Health & Suicide'},
    {'Variable Name': 'q106', 'Meaning': 'Difficulty with memory or decisions', 'Tag': 'Mental Health & Suicide'},               #because of a physical, mental, or emotional problem

    # Tobacco/Nicotine Use
    {'Variable Name': 'q31', 'Meaning': 'Ever smoked cigarette', 'Tag': 'Tobacco/Nicotine Use'},
    {'Variable Name': 'q32', 'Meaning': 'Age first smoked cigarette', 'Tag': 'Tobacco/Nicotine Use'},
    {'Variable Name': 'q33', 'Meaning': 'Days smoked cigarettes (30 days)', 'Tag': 'Tobacco/Nicotine Use'},
    {'Variable Name': 'q34', 'Meaning': 'Cigs/day on days smoked', 'Tag': 'Tobacco/Nicotine Use'},
    {'Variable Name': 'q35', 'Meaning': 'Ever used electronic vapor product', 'Tag': 'Tobacco/Nicotine Use'},
    {'Variable Name': 'q36', 'Meaning': 'Days used vapor product (30 days)', 'Tag': 'Tobacco/Nicotine Use'},
    {'Variable Name': 'q37', 'Meaning': 'How obtained vapor products', 'Tag': 'Tobacco/Nicotine Use'},
    {'Variable Name': 'q38', 'Meaning': 'Days used chewing tobacco (30 days)', 'Tag': 'Tobacco/Nicotine Use'},
    {'Variable Name': 'q39', 'Meaning': 'Days smoked cigars (30 days)', 'Tag': 'Tobacco/Nicotine Use'},
    {'Variable Name': 'q40', 'Meaning': 'Tried to quit all tobacco (12 months)', 'Tag': 'Tobacco/Nicotine Use'},

    # Alcohol Use
    {'Variable Name': 'q41', 'Meaning': 'Age first drank alcohol', 'Tag': 'Alcohol Use'},
    {'Variable Name': 'q42', 'Meaning': 'Days drank alcohol (30 days)', 'Tag': 'Alcohol Use'},
    {'Variable Name': 'q43', 'Meaning': 'Days had 4+/5+ drinks in row (30 days)', 'Tag': 'Alcohol Use'},
    {'Variable Name': 'q44', 'Meaning': 'Max drinks in a row (30 days)', 'Tag': 'Alcohol Use'},
    {'Variable Name': 'q45', 'Meaning': 'How obtained alcohol (30 days)', 'Tag': 'Alcohol Use'},
    {'Variable Name': 'q60', 'Meaning': 'Used alcohol/drugs before last sex', 'Tag': 'Alcohol Use'},

    # Drugs: Marijuana, Pain Meds, Other Illicit
    {'Variable Name': 'q46', 'Meaning': 'Times used marijuana (lifetime)', 'Tag': 'Other Drugs'},
    {'Variable Name': 'q47', 'Meaning': 'Age first tried marijuana', 'Tag': 'Other Drugs'},
    {'Variable Name': 'q48', 'Meaning': 'Times used marijuana (30 days)', 'Tag': 'Other Drugs'},
    {'Variable Name': 'q49/q92', 'Meaning': 'Times misused pain medicine', 'Tag': 'Other Drugs'},
    {'Variable Name': 'q50', 'Meaning': 'Times used cocaine', 'Tag': 'Other Drugs'},
    {'Variable Name': 'q51', 'Meaning': 'Times used inhalants', 'Tag': 'Other Drugs'},
    {'Variable Name': 'q52', 'Meaning': 'Times used heroin', 'Tag': 'Other Drugs'},
    {'Variable Name': 'q53', 'Meaning': 'Times used methamphetamines', 'Tag': 'Other Drugs'},
    {'Variable Name': 'q54', 'Meaning': 'Times used ecstasy', 'Tag': 'Other Drugs'},
    {'Variable Name': 'q55', 'Meaning': 'Times injected illegal drugs', 'Tag': 'Other Drugs'},
    {'Variable Name': 'q93', 'Meaning': 'Times used hallucinogens (lifetime)', 'Tag': 'Other Drugs'},

    # Sexual Activity & Identity
    {'Variable Name': 'q56', 'Meaning': 'Ever had sexual intercourse', 'Tag': 'Sexual Activity & Identity'},
    {'Variable Name': 'q57', 'Meaning': 'Age at first intercourse', 'Tag': 'Sexual Activity & Identity'},
    {'Variable Name': 'q58', 'Meaning': 'Lifetime sexual partners', 'Tag': 'Sexual Activity & Identity'},
    {'Variable Name': 'q59', 'Meaning': 'Sexual partners (past 3 months)', 'Tag': 'Sexual Activity & Identity'},
    {'Variable Name': 'q61', 'Meaning': 'Used condom at last sex', 'Tag': 'Sexual Activity & Identity'},
    {'Variable Name': 'q62', 'Meaning': 'Pregnancy prevention method at last sex', 'Tag': 'Sexual Activity & Identity'},
    {'Variable Name': 'q94', 'Meaning': 'Asked for verbal consent at last sexual encounter', 'Tag': 'Sexual Activity & Identity'},

    # Sex contact/Sexual/Gender Identity
    {'Variable Name': 'q63', 'Meaning': 'Sexual contact: who with (lifetime)', 'Tag': 'Sexual Contact'},
    {'Variable Name': 'q64', 'Meaning': 'Sexual identity', 'Tag': 'Sexual Identity'},
    {'Variable Name': 'q65', 'Meaning': 'Transgender identity', 'Tag': 'Gender Identity'},

    # Physical/Sexual Abuse & Parental Factors
    {'Variable Name': 'q19', 'Meaning': 'Physically Forced to have sex', 'Tag': 'Abuse & Parental Factors'},
    {'Variable Name': 'q20', 'Meaning': 'Times Forced to do sexual things', 'Tag': 'Abuse & Parental Factors'},
    {'Variable Name': 'q21', 'Meaning': 'Forced by partner to do sexual things', 'Tag': 'Abuse & Parental Factors'},
    {'Variable Name': 'q22', 'Meaning': 'Physically hurt by partner', 'Tag': 'Abuse & Parental Factors'},
    {'Variable Name': 'q88', 'Meaning': 'Sexual abuse by older adult', 'Tag': 'Abuse & Parental Factors'},
    {'Variable Name': 'q90', 'Meaning': 'Parent/adult physical abuse (lifetime)', 'Tag': 'Abuse & Parental Factors'},
    {'Variable Name': 'q89', 'Meaning': 'Parent/adult insulted/put down (lifetime)', 'Tag': 'Abuse & Parental Factors'},
    {'Variable Name': 'q91', 'Meaning': 'Parents/adults fought each other (lifetime)', 'Tag': 'Abuse & Parental Factors'},
    {'Variable Name': 'q99', 'Meaning': 'Adult met basic needs (lifetime)', 'Tag': 'Abuse & Parental Factors'},
    {'Variable Name': 'q100', 'Meaning': 'Lived with addict parent/guardian', 'Tag': 'Abuse & Parental Factors'},
    {'Variable Name': 'q101', 'Meaning': 'Lived with mentally ill parent/guardian', 'Tag': 'Abuse & Parental Factors'},
    {'Variable Name': 'q102', 'Meaning': 'Separated from parent because of incarceration', 'Tag': 'Abuse & Parental Factors'},
    {'Variable Name': 'q104', 'Meaning': 'An Adult knows whereabouts', 'Tag': 'Abuse & Parental Factors'},

    # Academic and Social School Experience
    {'Variable Name': 'q87', 'Meaning': 'Grades in school (12 months)', 'Tag': 'Academic & School Experience'},
    {'Variable Name': 'q103', 'Meaning': 'Feel close to people at school', 'Tag': 'Academic & School Experience'},

    # Physical/Mental disability Injury & Social/Medical
    {'Variable Name': 'q79', 'Meaning': 'Concussions from sports/activity (12 months)', 'Tag': 'Substance Use Injury & Social/Medical'},
    {'Variable Name': 'q98', 'Meaning': 'Times had sunburn (12 months)', 'Tag': 'Substance Use Injury & Social/Medical'},
    {'Variable Name': 'q80', 'Meaning': 'Social media use frequency', 'Tag': 'Substance Use Injury & Social/Medical'},
    {'Variable Name': 'q81', 'Meaning': 'Tested for HIV (ever)', 'Tag': 'Substance Use Injury & Social/Medical'},
    {'Variable Name': 'q82', 'Meaning': 'STD test (12 months)', 'Tag': 'Substance Use Injury & Social/Medical'},
    {'Variable Name': 'q83', 'Meaning': 'Last dental visit', 'Tag': 'Substance Use Injury & Social/Medical'},


    # Weight, Diet, Sleep
    {'Variable Name': 'q66', 'Meaning': 'Describe your weight', 'Tag': 'Weight, Diet, Sleep'},
    {'Variable Name': 'q67', 'Meaning': 'Current weight control effort', 'Tag': 'Weight, Diet, Sleep'},
    {'Variable Name': 'q68', 'Meaning': 'Drank 100% fruit juice (7 days)', 'Tag': 'Weight, Diet, Sleep'},
    {'Variable Name': 'q69', 'Meaning': 'Ate fruit (7 days)', 'Tag': 'Weight, Diet, Sleep'},
    {'Variable Name': 'q70', 'Meaning': 'Ate green salad (7 days)', 'Tag': 'Weight, Diet, Sleep'},
    {'Variable Name': 'q71', 'Meaning': 'Ate potatoes (7 days)', 'Tag': 'Weight, Diet, Sleep'},
    {'Variable Name': 'q72', 'Meaning': 'Ate carrots (7 days)', 'Tag': 'Weight, Diet, Sleep'},
    {'Variable Name': 'q73', 'Meaning': 'Ate vegetables (7 days)', 'Tag': 'Weight, Diet, Sleep'},
    {'Variable Name': 'q74', 'Meaning': 'Drank soda/pop (7 days)', 'Tag': 'Weight, Diet, Sleep'},
    {'Variable Name': 'q75', 'Meaning': 'Days ate breakfast (7 days)', 'Tag': 'Weight, Diet, Sleep'},
    {'Variable Name': 'q85', 'Meaning': 'Sleep hours (school night)', 'Tag': 'Weight, Diet, Sleep'},
    {'Variable Name': 'q86', 'Meaning': 'Slept usually where (30 days)', 'Tag': 'Weight, Diet, Sleep'},

    # Physical Activity
    {'Variable Name': 'q76', 'Meaning': 'Days active 60+ min (7 days)', 'Tag': 'Physical Activity'},
    {'Variable Name': 'q77', 'Meaning': 'PE class days (week)', 'Tag': 'Physical Activity'},
    {'Variable Name': 'q78', 'Meaning': 'Sports teams played (12 months)', 'Tag': 'Physical Activity'},

    # Communication and Language Skills
    {'Variable Name': 'q107', 'Meaning': 'English speaking ability', 'Tag': 'Communication and Language Skills'},
]

df = pd.DataFrame(data)
print(df)




```
Sexual Identity and sex of sexual contact are two components of Sexual Orientation. Third componenent sexual attraction is not
ascertained on the standard or the national YRBS questionnaire. Each of these three components are separate and independent constructs.
There could be discordance between sexual identity and sex of sexual contact.


I need to make sure which sexual minority variable should be used for my research question
```





Creating Sexual Minority variable

In [ ]:
#1. Sexual Minority from sexual identity, q64
# Rename column
raw_data = raw_data.rename(columns={'q64': 'sex_iden'})

sex_iden_labels = {
    1: 'Heterosexual (straight)',
    2: 'Gay or lesbian',
    3: 'Bisexual',
    4: 'Other way (self-described)',
    5: 'Questioning/not sure',
    6: 'Don’t know/missing'
}

#adding a mapping variable
raw_data['sex_iden_label'] = raw_data['sex_iden'].map(sex_iden_labels)

In [ ]:
#2. Sexual Minority from Sex and sex of sexual contacts


#Assigning sexual contact variable values


sex_labels = {1: 'male', 2: 'female'}
sexcontact_labels = {
    1: 'opposite sex only',
    2: 'same sex only',
    3: 'both sexes',
    4: 'no sexual contact'
}


def assign_sexcontact(row):
  if row['q2'] == '1':
      sex=2                  #female
      if row['q63']=='1':    #no sexual contact
        sexcontact=4
      elif row['q63']=='2':  #females/same
        sexcontact=2
      elif row['q63']=='3':  #males/opposite
        sexcontact=1
      elif row['q63']=='4':  #both males and females
        sexcontact=3
      else:
        sexcontact=None
  elif row['q2'] == '2':
      sex=1                   #male
      if row['q63']=='1':
        sexcontact=4          #no sexual contact
      elif row['q63']=='2':
        sexcontact=1          #females/opposite
      elif row['q63']=='3':
        sexcontact=2          #males/same
      elif row['q63']=='4':
        sexcontact=3          #both males and females
      else:
        sex = None
        sexcontact=None
  else:
        sex = None
        sexcontact= None
  return pd.Series({'sex':sex, 'sexual contact'=sexcontact})

raw_data[['sex', 'sexcontact']] = raw_data.apply(assign_sexcontact, axis=1)

# raw_data['sex_label'] = raw_data['sex'].map(sex_labels)
# raw_data['sexcontact_label'] = raw_data['sexcontact'].map(sexcontact_labels)



